### ISSUE TITLE
feature request/security: avoid container running root

### ISSUE THREAD

**Issue by vmaillot** [apache/apisix-docker#326]
Hey there,

In advance, thanks for your time and help.

# Context

I was trying to make Apisix run in a OpenShift cluster but without having root running the container. To do so I am using the official upstream Helm [chart](https://github.com/apache/apisix-helm-chart). I was able to run everything but the Apisix deployment without root.
A work around is to use the [SCC](https://docs.openshift.com/container-platform/4.10/authentication/managing-security-context-constraints.html) as `anyuid` on the service account running the application to let the application run as user root (0). Obviously in a perfect world we do not want that.

That is why I am asking if that would be possible to make the application run as a non-root user. When trying this solution with the help of the Helm chart, and without setting up the SCC I am getting this error below:

```
/usr/local/openresty/luajit/bin/luajit ./apisix/cli/apisix.lua init failed to update nginx.conf: failed to open file: /usr/local/apisi/conf/nginx.conf, error info:/usr/local/apisix/conf/nginx.conf: Permission denied
```

The error makes total sense, as the generated nginx configuration has those permissions:

```
bash-5.1# ls -la /usr/local/apisix/conf/nginx.conf
-rw-r--r--    1 root     root         11422 Jul  5 10:49 /usr/local/apisix/conf/nginx.conf
```

# Solutions

The first easy solution would be to add write access to the group root on this file as every container is started with the root group (0) on OpenShift. The nginx configuration seems to be generated somehow, maybe it would be doable to add the write permission to the group on the fly?

The second one is a bit more tricky and might require more work but would simply follows the [Best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user). Again maybe we have a application limitation, but I think that should be feasible!

Please let me know if something is unclear or not understood here.

Thanks for evaluating this proposition of a new feature to gain a bit of flexibility and security while running Apisix.

**Comment by tokers** [comment]
> The first easy solution would be to add write access to the group root on this file as every container is started with the root group (0) on OpenShift. The nginx configuration seems to be generated somehow, maybe it would be doable to add the write permission to the group on the fly?

Yes, the `nginx.conf` is generated when APISIX just starts. We may add the write and execute permission for the path `/usr/local/apisix/conf`.

**Comment by vmaillot** [comment]
> Yes, the `nginx.conf` is generated when APISIX just starts. We may add the write and execute permission for the path `/usr/local/apisix/conf`.

Right, your suggestion is nice and it would be awesome if you could implement this feature. Thanks for your effort here!

**Comment by tzssangglass** [comment]
> Yes, the `nginx.conf` is generated when APISIX just starts. We may add the write and execute permission for the path `/usr/local/apisix/conf`.

we can do this in Dockerfile?

**Comment by tokers** [comment]
> > Yes, the `nginx.conf` is generated when APISIX just starts. We may add the write and execute permission for the path `/usr/local/apisix/conf`.
> 
> 
> 
> we can do this in Dockerfile?

I think so.

**Comment by vmaillot** [comment]
@tzssangglass @tokers any news?

**Comment by tzssangglass** [comment]
someone work on this:https://github.com/apache/apisix-docker/search?q=root+is%3Apr&type=issues, but no one seems to be pushing it to master eventually.

**Comment by vmaillot** [comment]
Any new status on that?

**Comment by vmaillot** [comment]
ping :}

**Comment by tokers** [comment]
> someone work on this:https://github.com/apache/apisix-docker/search?q=root+is%3Apr&type=issues, but no one seems to be pushing it to master eventually.

@vmaillot Is this OK for you?

**Comment by vmaillot** [comment]
> > someone work on this:https://github.com/apache/apisix-docker/search?q=root+is%3Apr&type=issues, but no one seems to be pushing it to master eventually.
> 
> @vmaillot Is this OK for you?

well, can't this be pushed to master somehow?

**Comment by Hazel6869** [comment]
> 
 hi, i will do it  @vmaillot

**Comment by vmaillot** [comment]
> > 
> 
> hi, i will do it @vmaillot

hey @Hazel6869 thanks a lot! it is much appreciated

**Comment by vmaillot** [comment]
@Hazel6869 any progress on your side regarding #371?

**Comment by Hazel6869** [comment]
@vmaillot  I'm testing.  sorry～

**Comment by vmaillot** [comment]
> @vmaillot I'm testing. sorry～ I want to know which image is you used, ![image](https://user-images.githubusercontent.com/30279880/204233252-195c8575-13d2-4cc2-bf8f-89e60c97dd17.png)

hey @Hazel6869 sorry for the late reply, but i would suggest to go with alpine : D lightweight ftw!

**Comment by soulbird** [comment]
APISIX has stopped maintaining the alpine image, it is recommended that you use debian.

**Comment by Hazel6869** [comment]
@vmaillot sorry。APISIX has stopped maintaining the alpine image, it is recommended that you use debian.  how do you think.

**Comment by vmaillot** [comment]
> @vmaillot sorry, APISIX has stopped maintaining the alpine image, it is recommended that you use debian. how do you think.

@Hazel6869 debian would totally fine! thanks for checking

**Comment by Hazel6869** [comment]
@vmaillot OK,  I will use debian,  I have time today, I will fix it.

**Comment by fengxsong** [comment]
Does this change need to be applied to centos based image? and also, this may required some more docs, non root user can not bind ports under 1024, unless use `securityContext.runAsUser` to override the default user.

**Comment by ankworld** [comment]
Finally, I found cause that I can't bind ports 80 and 443. Release 3.1.0 is the first release that contain this change, right? This affect helm chart too. Have to add securityContext.runAsUser=0 (run as root).

**Comment by tokers** [comment]
> Finally, I found cause that I can't bind ports 80 and 443. Release 3.1.0 is the first release that contain this change, right? This affect helm chart too. Have to add securityContext.runAsUser=0 (run as root).

@Hazel6869  Please take a look when you have time.

**Comment by vmaillot** [comment]
@Hazel6869 it looks like we still have permissions issues with the `nginx.conf` file:

```
/usr/local/openresty//luajit/bin/luajit ./apisix/cli/apisix.lua init
failed to update nginx.conf: failed to open file: /usr/local/apisix/conf/nginx.conf, error info:/usr/local/apisix/conf/nginx.conf: Permission denied
```

The running pod is using the latest `apisix 3.1.0-debian` image (with hash `apache/apisix@sha256:5901f303007d1853136ea6d95bca874a838ca90f4047bb984ff7a82c9575228c`).

From an Helm chart point of view, why did you guys reverted to use `2.15.1-debian` [image](https://github.com/apache/apisix-helm-chart/blob/apisix-0.13.0/charts/apisix/values.yaml#L84)?

Thanks for your time and help!

**Comment by ankworld** [comment]
Chart 1.x is for Apisix v3.
Chart 0.x is for Apisix V2.

**Comment by mkocsis-tg** [comment]
@ankworld The helm chart version 0.12.3 contains / referencing the apisix 3.1.0-debian image. So the question above was valid. 
https://github.com/apache/apisix-helm-chart/blob/apisix-0.12.3/charts/apisix/values.yaml#L84

This was fixed in the chart 0.13.0:
https://github.com/apache/apisix-helm-chart/blob/apisix-0.13.0/charts/apisix/values.yaml#L84

**Comment by Hazel6869** [comment]
> From an Helm chart point of view, why did you guys reverted to use `2.15.1-debian` [image](https://github.com/apache/apisix-helm-chart/blob/apisix-0.13.0/charts/apisix/values.yaml#L84)?
> 
> Thanks for your time and help!

I don't understand, does the current change meet your needs?
